### PR TITLE
Add passing specs for #valid_move? for Knight

### PR DESCRIPTION
**Pull request by radiofreemattd**
This is largely modeled after the King.

### CODE REVIEW

File: app/models/knight.rb [hunk 1]
```diff
@@ -1,2 +1,17 @@
...
+    return false if out_of_bounds?(destination_x, destination_y)
+    return false unless x_and_y_are_offset?(destination_x, destination_y)
+    true
+  end
+
+  private
+
+  def x_and_y_are_offset?(destination_x, destination_y)
+    ((destination_x - position_x == XY_MOVE_FACTOR) && (destination_y - position_y == ALT_XY_MOVE_FACTOR)) ||
+      ((destination_y - position_y == XY_MOVE_FACTOR) && (destination_x - position_x == ALT_XY_MOVE_FACTOR))
```
Comment (jonmast) [inline app/models/knight.rb]: I think you need to get the absolute value of the difference here(like you did in king).
I think this test will fail.
```ruby
it 'returns true if the x and y move components are -1 and 2, respectively' do
 expect(knight.valid_move?(3, 6)).to eq true
end
```

Comment (radiofreemattd) [inline app/models/knight.rb]: Totally see that...what if it were to look like this?

```ruby
class Knight < ChessPiece
  XY_MOVE_FACTOR = 1
  ALT_XY_MOVE_FACTOR = 2

  def valid_move?(destination_x, destination_y)
    return false if out_of_bounds?(destination_x, destination_y)
    return false unless x_and_y_are_offset?(destination_x, destination_y)
    true
  end

  private

  def x_and_y_are_offset?(destination_x, destination_y)
    ((diff_in_x(destination_x) == XY_MOVE_FACTOR) && (diff_in_y(destination_y) == ALT_XY_MOVE_FACTOR)) ||
      ((diff_in_y(destination_y) == XY_MOVE_FACTOR) && (diff_in_x(destination_x) == ALT_XY_MOVE_FACTOR))
  end

  def diff_in_x(destination_x)
    (destination_x - position_x).abs
  end

  def diff_in_y(destination_y)
    (destination_y - position_y).abs
  end
end
```

File: app/models/knight.rb [hunk 2]
```diff
@@ -1,2 +1,17 @@
 class Knight < ChessPiece
+  XY_MOVE_FACTOR = 1
+  ALT_XY_MOVE_FACTOR = 2
```
Comment (mscoutermarsh) [inline app/models/knight.rb]: is there a more descriptive name for these constants?

I'm not sure I really know what they mean.

File: app/models/knight.rb [hunk 3]
```diff
@@ -1,2 +1,17 @@
...
+
+  def valid_move?(destination_x, destination_y)
+    return false if out_of_bounds?(destination_x, destination_y)
+    return false unless x_and_y_are_offset?(destination_x, destination_y)
+    true
```
Comment (mscoutermarsh) [inline app/models/knight.rb]:
```Ruby
return false if out_of_bounds?(destination_x, destination_y)
x_and_y_are_offset?(destination_x, destination_y)
```

File: app/models/knight.rb (hunk 2 above)
Comment (radiofreemattd) [inline app/models/knight.rb]: Probably...I struggled myself. Basically, I am trying to express that both x and y can either be 1 or 2 - so I modeled this concept of the `XY_MOVE_FACTOR` as being 1 and the alternative as being 2.

It sucks and I am totally open to other concepts... :)

File: app/models/knight.rb (hunk 1 above)
Comment (mscoutermarsh) [inline app/models/knight.rb]: How about.
```Ruby
# A Knight can move up 1 and over 2. Or up 2 and over 1
VALID_MOVE_DISTANCES = [[1, 2], [2, 1]]

## Then you could do this...
def valid_move?(destination_x, destination_y)
    return false if out_of_bounds?(destination_x, destination_y)
    VALID_MOVE_DISTANCES.include? [diff_in_x(destination_x), diff_in_y(destination_y)]
end
```

(I didn't test this, but I think will work) :smile:

**Pull request merged**